[RFC] Kernel TLS socket API

borisPis opened the pull request:
(DO NOT SUBMIT. Request for comments only.)

I would like to introduce you to the new kernel API for TLS transmit-side data-path, and open a discussion regarding its support in OpenSSL.

This is currently a V2 patch series in Linux net-next, and it is stabilizing.
Dave has been working on this for a while [[1](https://netdevconf.org/1.2/session.html?dave-watson)][[2](https://www.spinics.net/lists/linux-crypto/msg25953.html)], and Aviad, Ilya and myself have also been working on TLS transmit-side crypto offload to a NIC [[3]](https://netdevconf.org/1.2/session.html?boris-pismenny)[[4]](https://www.spinics.net/lists/netdev/msg427122.html).
Similar work is also being done in FreeBSD by Netflix [[5]](https://people.freebsd.org/~rrs/asiabsd_2015_tls.pdf)[[6]](https://people.freebsd.org/~rrs/asiabsd_tls_improved.pdf).

Placing the data-path of TLS in the kernel has several advantages. First, an efficient TLS sendfile system call implementation is possible with such a socket. Second, data from userspace could be encrypted when passing the user-kernel boundary, reducing the overhead of existing implementations that require at least encrypting the data and then copying from user to kernel.
Third, TLS data-path crypto offload to hardware requires a kernel infrastructure for TLS.

To sum up, a new kernel socket option is introduced to add support for the data-path of TLS and it will soon be common in operating systems.

For anyone who is interested, here is the kernel code:
https://github.com/ktls/net_next_ktls/commits/ktls_v2

The user API is as follows: (also see [[7]](https://www.spinics.net/lists/linux-crypto/msg25957.html))

Creating a TLS connection
-------------------------
First create a new TCP socket and set the TLS ULP.

```
  sock = socket(AF_INET, SOCK_STREAM, 0);
  setsockopt(sock, SOL_TCP, TCP_ULP, "tls", sizeof("tls"));
```

Setting the TLS ULP allows us to set/get TLS socket options. 
Currently only the symmetric encryption is handled in the kernel. Patch 2 sets the TLS ULP in BIO_new_socket.

All the parameters required to move the data-path to the kernel are available after processing the change cipher state message during the handshake.
Patch 4 adds code similar to the one below in tls1_change_cipher_state.

There is a separate socket option for moving the transmit and the receive into the kernel. Current code, supports only the transmit side, TLS1.2 and AES-GCM128. In the future, this will be extended.

Below is the code that sets the socket option which moves the transmit data-path into the kernel. Patch 1 adds a temporary netinet/tcp.h file which includes these structures. Do you know if netinet/tcp.h is the right place for these?

```
 /* From linux/tls.h */
 struct tls_crypto_info {
         unsigned short version;
         unsigned short cipher_type;
 };

  struct tls12_crypto_info_aes_gcm_128 {
          struct tls_crypto_info info;
          unsigned char iv[TLS_CIPHER_AES_GCM_128_IV_SIZE];
          unsigned char key[TLS_CIPHER_AES_GCM_128_KEY_SIZE];
          unsigned char salt[TLS_CIPHER_AES_GCM_128_SALT_SIZE];
          unsigned char rec_seq[TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE];
  };

  struct tls12_crypto_info_aes_gcm_128 crypto_info;

  crypto_info.info.version = TLS_1_2_VERSION;
  crypto_info.info.cipher_type = TLS_CIPHER_AES_GCM_128;
  memcpy(crypto_info.iv, iv_write, TLS_CIPHER_AES_GCM_128_IV_SIZE);
  memcpy(crypto_info.rec_seq, seq_number_write, TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
  memcpy(crypto_info.key, cipher_key_write, TLS_CIPHER_AES_GCM_128_KEY_SIZE);
  memcpy(crypto_info.salt, implicit_iv_write, TLS_CIPHER_AES_GCM_128_SALT_SIZE);

  setsockopt(sock, SOL_TLS, TLS_TX, &crypto_info, sizeof(crypto_info));
```

Sending TLS application data
----------------------------

After setting the TLS_TX socket option all application data sent over this socket is encrypted using TLS and the parameters provided in the socket option.

For example, an application can send an encrypted hello world record as follows:

```
  const char *msg = "hello world\n";
  send(sock, msg, strlen(msg));

```
send() data is directly encrypted from the userspace buffer provided to the encrypted kernel send buffer if possible.

The sendfile system call will send the file's data over TLS records of maximum length (2^14).

```
  file = open(filename, O_RDONLY);
  fstat(file, &stat);
  sendfile(sock, file, &offset, stat.st_size);
```

TLS records are created and sent after each send() call, unless MSG_MORE is passed.  MSG_MORE will delay creation of a record until MSG_MORE is not passed, or the maximum record size is reached or an alert record needs to be sent.

The kernel will need to allocate a buffer for the encrypted data.
This buffer is allocated at the time send() is called, such that either the entire send() call will return -ENOMEM (or block waiting for memory), or the encryption will always succeed.  If send() returns -ENOMEM and some data was left on the socket buffer from a previous call using MSG_MORE, the MSG_MORE data is left on the socket buffer.

Send TLS control messages
-------------------------

Other than application data, TLS has control messages such as alert messages (record type 21) and handshake messages (record type 22), etc. These messages can be sent over the socket by providing the TLS record type via a CMSG.
Patch 2 adds the following function (in bss_sock.c) that sends @data of @length bytes using a record of type @record_type.
```
/* send TLS control message using record_type */
  static int klts_send_ctrl_message(int sock, unsigned char record_type,
                                  void *data, size_t length)
  {
        struct msghdr msg = {0};
        int cmsg_len = sizeof(record_type);
        struct cmsghdr *cmsg;
        char buf[CMSG_SPACE(cmsg_len)];
        struct iovec msg_iov;   /* Vector of data to send/receive into.  */

        msg.msg_control = buf;
        msg.msg_controllen = sizeof(buf);
        cmsg = CMSG_FIRSTHDR(&msg);
        cmsg->cmsg_level = SOL_TLS;
        cmsg->cmsg_type = TLS_SET_RECORD_TYPE;
        cmsg->cmsg_len = CMSG_LEN(cmsg_len);
        *CMSG_DATA(cmsg) = record_type;
        msg.msg_controllen = cmsg->cmsg_len;

        msg_iov.iov_base = data;
        msg_iov.iov_len = length;
        msg.msg_iov = &msg_iov;
        msg.msg_iovlen = 1;

        return sendmsg(sock, &msg, 0);
  }
```

Control message data should be provided unencrypted, and will be encrypted by the kernel.  At a high level, the kernel TLS ULP is a replacement for the record layer of a userspace TLS library.

References:
[1] - https://netdevconf.org/1.2/session.html?dave-watson
[2] - https://www.spinics.net/lists/linux-crypto/msg25953.html
[3] - https://netdevconf.org/1.2/session.html?boris-pismenny
[4] - https://www.spinics.net/lists/netdev/msg427122.html
[5] - https://people.freebsd.org/~rrs/asiabsd_2015_tls.pdf
[6] - https://people.freebsd.org/~rrs/asiabsd_tls_improved.pdf
[7] - https://www.spinics.net/lists/linux-crypto/msg25957.html

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

mattcaswell commented on crypto/bio/bss_sock.c:
```diff
@@ -103,12 +118,54 @@ static int sock_read(BIO *b, char *out, int outl)
+    if (BIO_should_offload_tx_ctrl_msg_flag(b)) {
+        unsigned char record_type = (unsigned char)b->ptr;
+
+#ifdef SSL_DEBUG
```
Not sure this define really belongs in libcrypto (and the same for other instances of this in this file)

mattcaswell commented on crypto/bio/bss_sock.c:
```diff
@@ -103,12 +118,54 @@ static int sock_read(BIO *b, char *out, int outl)
+#endif
+        BIO_clear_offload_tx_ctrl_msg_flag(b);
+        ret = send_ctrl_message(b->num, record_type, in, inl);
+        if (ret >= 0) {
```
Style: No {} for single line if.

mattcaswell commented on e_os.h:
```diff
@@ -423,7 +423,7 @@ struct servent *PASCAL getservbyname(const char *, const char *);
 #    include <netinet/in.h>
 #    include <arpa/inet.h>
-#    include <netinet/tcp.h>
+#    include "netinet/tcp.h" // Add TLS stuff here..
```
This comment style is not allowed. Only /* */

bernd-edlinger replied:
and why include ".." instead of <..> ?

mattcaswell commented on include/openssl/bio.h:
```diff
@@ -187,6 +199,18 @@ void BIO_clear_flags(BIO *b, int flags);
+# define BIO_should_offload_tx_ctrl_msg_flag(b) \
+    BIO_test_flags(b, (BIO_FLAGS_OFFLOAD_TX_CTRL_MSG))
+# define BIO_clear_offload_tx_ctrl_msg_flag(b) \
+    BIO_clear_flags(b, (BIO_FLAGS_OFFLOAD_TX_CTRL_MSG))
```
Do these macros belong in the public header? i.e. are they part of the API?

mattcaswell commented on ssl/record/rec_layer_s3.c:
```diff
@@ -759,24 +759,31 @@ int do_ssl3_write(SSL *s, int type, const unsigned char *buf,
-        SSL3_RECORD_set_type(&wr[j], type);
+        if (BIO_get_offload_tx(s->wbio)) {
+            /* IV will be generated by the kernel if offload is used */
+            eivlen = 0;
```
It seems more appropriate for this to be set in the section of code above where eivlen is originally set.

mattcaswell commented on ssl/record/rec_layer_s3.c:
```diff
@@ -759,24 +759,31 @@ int do_ssl3_write(SSL *s, int type, const unsigned char *buf,
-            *(outbuf[j]++) = 0x1;
-        else
-            *(outbuf[j]++) = s->version & 0xff;
+            SSL3_RECORD_set_type(&wr[j], type);
```
This seems to be done in both branches of the if, so perhaps this is better outside the if

mattcaswell commented on ssl/record/rec_layer_s3.c:
```diff
@@ -759,24 +759,31 @@ int do_ssl3_write(SSL *s, int type, const unsigned char *buf,
+            SSL3_RECORD_set_type(&wr[j], type);
+        } else {
+            /* write the header */
+            *(outbuf[j]++) = type & 0xff;
```
We seem to do this work twice. Once here if not offloading and the target buffer is outbuf[j]. Then we do it again later if we are offloading but that time the target buffer is the dummy "header". It seems more sensible to just do this once, but change the target buffer as appropriate.

mattcaswell commented on ssl/record/rec_layer_s3.c:
```diff
@@ -913,10 +947,21 @@ int ssl3_write_pending(SSL *s, int type, const unsigned char *buf,
+             */
+            if (BIO_get_offload_tx(s->wbio) &&
+                    type != SSL3_RT_APPLICATION_DATA) {
+                (void)BIO_flush(s->wbio);
```
So does that mean if multiple reneg handshake messages are all sent one after the other they are all going to end up in separate TCP packets rather than a single one?

mattcaswell commented on ssl/t1_enc.c:
```diff
@@ -314,6 +323,85 @@ int tls1_change_cipher_state(SSL *s, int which)
+    /* This is not the IV that is used by OpenSSL, but rather the sequence
+     * number. We should actually provide both iv and sequence.
+     */
+    EVP_CIPHER_CTX_ctrl(dd, EVP_CTRL_GCM_GET_IV, 12, geniv);
```
We should probably have a #define for the value 12 here...although that's not specific to this change

mattcaswell commented on ssl/t1_enc.c:
```diff
@@ -314,6 +323,85 @@ int tls1_change_cipher_state(SSL *s, int which)
+     * number. We should actually provide both iv and sequence.
+     */
+    EVP_CIPHER_CTX_ctrl(dd, EVP_CTRL_GCM_GET_IV, 12, geniv);
+    memcpy(crypto_info.iv, geniv + 4, TLS_CIPHER_AES_GCM_128_IV_SIZE);
```
What's the magic number 4 here? Should there be a define for it?

mattcaswell commented on ssl/t1_enc.c:
```diff
@@ -314,6 +323,85 @@ int tls1_change_cipher_state(SSL *s, int which)
+            s->msg_callback(2, s->version, wh, iv, k, s, s->msg_callback_arg);
+        }
+    }
+#endif
```
This whole block of code looks like a merge error. This code was removed from the codebase recently. You seem to be adding it back in.

borisPis commented:
Hi mattcaswell,

Thank you very much for reviewing the code and sorry for the late response.

Since these patches have been posted the Kernel TLS API has been accepted into the mainline tree and this API is available since Linux 4.13. 

I've refactored the changes and rebased them over the OpenSSL 1.1.0 branch. Due to the many changes required by the rebase. I've opened a separate pull request for it here: https://github.com/openssl/openssl/pull/5253

Regarding your questions:
- it is not possible to control the fragment length through the API at the moment. Support for this will be added in the future.
- Eventually, the asynchronous crypto engines provided by the afalg engine could be used by the kernel to perform crypto when copying data from user space. At the moment the kernel supports only symmetric crypto. Overall, I see no reason why anyone would want to use afalg directly from userspace after this change.
